README.md
# 一个文本转视频的工具

这个工具可以将一段文本转换为视频，并保存到指定的本地，初衷是想实现小说的可视化阅读功能。

效果图如下：

<img src="https://github.com/bravekingzhang/text2viedo/blob/main/demos/4301686560452_.pic.jpg" alt="效果图" style="width: 100%;" />

## 实现原理

- 将文本进行分段，现在没有想到好的办法，就是通过标点符号句号分段，分成一个个的句子
- 通过句子生成图片，生成声音，图片开源的有很多，本方案采用 stable-diffusion，语言转文字使用 edge-tts
- 在通过 opencv 将图片合并为视频，目前输出 mp4 格式的视频，句子作为字母贴到视频内容的底部区域。
- 音频是一个有时间概念的东西，恰好可以通过音频控制一张画面的播放时长
- 在通过 ffmpeg 将音频合并到原始视频中。

最终，一个有画面，有字幕，有声音的视频就出现了，咱们实现了一个 `文本转视频`。

# 本地开发

## 安装依赖

开发时，需要安装的环境是 `macOS` `python 3.10.12`，其他环境可能存在兼容性问题

pip install -r requirements.txt

## 生成 huggingface api key

token 申请地址：https://huggingface.co/settings/tokens

因为，该项目中使用了 huggingface 上的开源文生图模型生成图片，中文生成图片效果不大好，因此，本项目对中文进行了翻译，感谢有道，直接使用有道翻译，比较方便。翻译后，生成图的质量有一定的提高。

token 可以写入到 .env 文件里面

### 如果使用的 pollinations-ai ，则不填写 token 就 ok

## 安装 ffmpeg

因为视频合成声音需要

## 开始使用

```python
python3.10 app.py
http://127.0.0.1:5000/
```

## 创作不易，支持开源

如果你觉得对你有用，就请我一杯咖啡。

<img src="https://github.com/bravekingzhang/utools-code2flow-official/blob/main/shoukuanma.png" alt="收款码" style="width: 40%;" />

## 加群交流

<img src="https://github.com/bravekingzhang/flutter_chat_box/blob/main/artificial/qunliao.pic.jpg"  style="width: 50%;">

添加我，注明自己 github 账号名，入群的目的，加入群聊，一起学习交流。

# License: MIT

本项目采用 MIT 许可证授权。


CHANGELOG.MD
# 增加了 pollinations-ai 模型

不用 huggingface 模型也可以生成图片啦


app.py
from flask import Flask, request, render_template, jsonify, send_from_directory

from text_to_video import convertTextToVideo

from configs import models

app = Flask(__name__)


@app.route('/', methods=['GET'])
def index():
    return render_template('index.html')

@app.route('/convert', methods=['POST'])
def convert_text_to_video():
    # 获取请求中的文本数据
    text = request.form.get('text')
    model = request.form.get('model')
    # 在此处调用您的文本转视频的代码，并生成视频文件
    # videos/1686290882-anything-v4.0.mp4
    video_path = convertTextToVideo(validate_model(model), text)
    return jsonify({'video_path': video_path})


@app.route('/models', methods=['GET'])
def get_available_models():
    return jsonify(models)

@app.route('/videos/<path:filename>')
def get_video(filename):
    return send_from_directory('videos', filename)

def validate_model(model):
    if model in models:
        return model
    else:
        return models[0]




if __name__ == '__main__':
    app.run()


main.py
import time
import requests
import json
import cv2
import os
import textwrap
from dotenv import load_dotenv
import numpy as np
import subprocess
import re

from add_text_to_image import add_text_to_image
from translate import translate_to_english

models = ["stabilityai/stable-diffusion-2-1", "andite/anything-v4.0"]

# 尝试加载线上环境变量文件
load_dotenv('.env', override=True)

# 尝试加载本地开发环境变量文件
load_dotenv('.local.env', override=True)

# 获取当前脚本所在的目录
current_directory = os.getcwd()

# 读取环境变量
api_token = os.getenv('API_TOKEN')

headers = {
    "Authorization": f"Bearer {api_token}",
    "Content-Type": "application/json"
}
def generateImage(model, prompt):
    body = {
        "inputs": translate_to_english(prompt)
    }
    r = requests.post("https://api-inference.huggingface.co/models/" + model,
                      data=json.dumps(body), headers=headers)
    # 将图片写入到 images 目录下，每个图片使用(时间戳+model).png 来命名
    timeStamp = str(int(time.time()))
    imagePath = "images/" + timeStamp + \
        "-" + model.split("/")[-1] + ".png"
    with open(imagePath, "wb") as f:
        f.write(r.content)
        f.close()
    voicePath = "voices/" + timeStamp + \
        "-" + model.split("/")[-1] + ".mp3"
    convert_text_to_speech(
        text=prompt, output_file=voicePath
    )


def convert_text_to_speech(text, output_file):
        # 指定输出目录
    output_directory = os.path.join(current_directory,"voices")
    # 创建输出目录（如果不存在）
    os.makedirs(output_directory, exist_ok=True)
    # 执行命令，并将工作目录设置为输出目录
    try:
        command = ['edge-tts', '--voice', 'zh-CN-XiaoyiNeural', '--text', text, '--write-media', output_file, '--write-subtitles', f'{output_file}.vtt']
        result = subprocess.run(command, cwd=current_directory, timeout=10)
        print(result)
        duration = get_duration_from_vtt(output_file + ".vtt")
        # 删除 无效音频 or 重新生成？
        if duration == 0.1:
            os.remove(output_file + ".vtt")
            os.remove(output_file)

    except subprocess.CalledProcessError as e:
        print("Command execution failed with return code:", e.returncode)
        print("Command output:", e.output)

def clear_folder(folder_path):
    """清空指定文件夹中的文件"""
    for filename in os.listdir(folder_path):
        file_path = os.path.join(folder_path, filename)
        if os.path.isfile(file_path):
            os.remove(file_path)





def split_sentences(text):
    pattern = r'[,.，。]'
    sentences = re.split(pattern, text)
    # 移除空白的句子
    sentences = [sentence.strip()
                 for sentence in sentences if sentence.strip()]
    return sentences
def convertTextToVideo(model, text):

    # 将文本段落进行分句
    sentences = split_sentences(text)

    # 清空 images 文件夹
    clear_folder("images")
    # 清空 voices 文件夹
    clear_folder("voices")

    # 为每个句子生成图片
    for sentence in sentences:
        if sentence.strip() != "":
            generateImage(model, sentence.strip())

    # 合成视频
    frame_width = 640
    frame_height = 480
    timeStamp = str(int(time.time()))
    output_video_path = "videos/" + timeStamp + \
        "-" + model.split("/")[-1] + ".mp4"
    output_video = cv2.VideoWriter(output_video_path, cv2.VideoWriter_fourcc(
        *'mp4v'), 30, (frame_width, frame_height))

    image_files = os.listdir('images')
    image_files.sort()

    for image_file in image_files:
        if image_file.endswith(".png"):

            text_color = (255, 255, 255)  # 白色文字
            background = (0, 0, 0,128)  # 黑色背景半透明
            image_path = "images/" + image_file
            draw_text = sentences[image_files.index(image_file)]
            add_text_to_image(draw_text, image_path,
                              text_color, background, padding=10)
            image = cv2.imread(image_path)
            resized_image = cv2.resize(image, (frame_width, frame_height))
            output_video.write(resized_image)
            # 添加停顿帧
            duration = get_duration_from_vtt(
                f"voices/{find_file_name_without_extension(image_file)}.mp3.vtt")
            print(duration)
            for _ in range(int(duration * 30)):
                output_video.write(resized_image)

    output_video.release()
    desc_output_video_path = "videos/" + timeStamp + \
        "-" + model.split("/")[-1] + ".withAudio.mp4"

    merge_audio_to_video("voices", output_video_path,
                         desc_output_video_path)

def find_file_name_without_extension(file_path):
    file_name = os.path.basename(file_path)
    file_name_without_extension = os.path.splitext(file_name)[0]
    return file_name_without_extension
def merge_audio_to_video(audio_directory, video_file, output_file):
    # 获取目录中的音频文件
    audio_files = [file for file in os.listdir(
        audio_directory) if file.endswith('.mp3')]

    if not audio_files:
        print("No audio files found in the directory.")
        return
    audio_files.sort()
    # 生成FFmpeg命令
    command = [
        'ffmpeg',
        '-i',
        video_file,
    ]

    # 添加音频文件参数
    for audio_file in audio_files:
        command.extend(['-i', audio_directory+'/'+audio_file])

    # 设置音频合并选项
    command.extend([
        '-filter_complex',
        ''.join([f'[{i+1}:0]' for i in range(len(audio_files))]) +
        f'concat=n={len(audio_files)}:v=0:a=1[outa]',
        '-map',
        '0:v',
        '-map',
        '[outa]',
        '-c:v',
        'copy',
        '-c:a',
        'aac',
        '-shortest',
        output_file
    ])

    # 执行FFmpeg命令
    result = subprocess.run(command,cwd=current_directory, timeout=300)
    print(result)

def get_duration_from_vtt(vtt_file):
    print(vtt_file)
    if not os.path.exists(vtt_file):
        return 0.1
    with open(vtt_file, 'r') as file:
        lines = file.readlines()

    total_duration = 0.1

    for line in lines:
        line = line.strip()
        if '-->' in line:
            start_time, end_time = line.split('-->')
            start_time = start_time.strip()
            end_time = end_time.strip()
            start_seconds = convert_time_to_seconds(start_time)
            end_seconds = convert_time_to_seconds(end_time)
            duration = end_seconds - start_seconds
            total_duration += duration

    return total_duration


def convert_time_to_seconds(time):
    hours, minutes, seconds = time.split(':')
    seconds, milliseconds = seconds.split('.')
    hours = int(hours)
    minutes = int(minutes)
    seconds = int(seconds)
    milliseconds = int(milliseconds)
    total_seconds = (hours * 3600) + (minutes * 60) + \
        seconds + (milliseconds / 1000)
    return total_seconds


if __name__ == '__main__':
   text_test= '''
   一个风和日丽的早上，我骑着自行车去学校，在路上遇到了彩虹，当时我的心情非常的愉快。
'''
   convertTextToVideo(models[0], text_test)


add_text_to_image.py
from PIL import Image, ImageDraw, ImageFont
import jieba

def add_text_to_image(text, image_path, text_color, background, padding=10, target_size=(640, 480), font_path="fonts/Hiragino Sans GB.ttc", font_size=16):
    # Open the image
    image = Image.open(image_path)

    # Resize the image to the target size while maintaining the aspect ratio
    image = image.resize(target_size)

    # Convert the image to RGBA mode
    image = image.convert("RGBA")

    # Create a new transparent image with the same size as the original image
    overlay = Image.new("RGBA", image.size, (0, 0, 0, 0))

    # Create a draw object for the overlay image
    draw = ImageDraw.Draw(overlay)

    # Load the font
    font = ImageFont.truetype(font_path, font_size)

    # Split the text into lines if it exceeds the available width
    lines = []

    # Split the text into individual Chinese characters
    words = [char for char in jieba.cut(text)]
    current_line = words[0]
    for word in words[1:]:
        if draw.textsize(current_line  + word, font=font)[0] <= target_size[0] - 2 * padding:
            current_line +=  word
        else:
            lines.append(current_line)
            current_line = word
    lines.append(current_line)

    # Calculate the height of the background rectangle based on the number of lines
    text_height = draw.textsize(lines[0], font=font)[1]

    if len(lines) == 1:
        box_height = text_height + padding * 2
    else:
        box_height = (text_height + padding) * len(lines) +  padding

    # Calculate the position of the background rectangle
    box_position = ((target_size[0] - draw.textsize(lines[0], font=font)
                    [0]) // 2 - padding, target_size[1] - box_height - padding)

    # Calculate the width of the background rectangle
    box_width = draw.textsize(lines[0], font=font)[0] + 2 * padding

    # Draw a rectangle with the specified background color and alpha
    draw.rectangle(
        (box_position, (box_position[0] + box_width, box_position[1] + box_height)), fill=background)

    # Calculate the starting y-position for drawing the text
    start_y = box_position[1] + padding

    # Draw the text on the overlay image
    for line in lines:
        text_width, text_height = draw.textsize(line, font=font)
        text_position = ((target_size[0] - text_width) // 2, start_y)
        print(line, text_position)
        draw.text(text_position, line, font=font, fill=text_color)
        start_y += text_height + padding

    # Paste the overlay image onto the original image using alpha composite
    image = Image.alpha_composite(image, overlay)

    # Convert the image back to RGB mode
    image = image.convert("RGB")

    # Save the resulting image
    image.save(image_path)


# # 示例用法
# text = "这是一段测试字幕。这段字幕可能会比较长，需要换行显示。如果字幕文本的宽度超过了目标图像的宽度，将会自动换行显示，并确保整体居中。"
# image_path = "images/1686290882-anything-v4.0.png"
# text_color = (255, 255, 255)  # 白色字体
# background = (0, 0, 0, 128)  # 半透明黑色底部浮层
# padding = 10
# target_size = (640, 480)
# font_path = "fonts/Hiragino Sans GB.ttc"
# font_size = 16

# add_text_to_image(text, image_path, text_color, background,
#                   padding, target_size, font_path, font_size)


text_to_video.py
import time
import requests
import json
import cv2
import os
import textwrap
from dotenv import load_dotenv
import numpy as np
import subprocess
import re

from add_text_to_image import add_text_to_image
from translate import translate_to_english



# 尝试加载线上环境变量文件
load_dotenv('.env', override=True)

# 尝试加载本地开发环境变量文件
load_dotenv('.local.env', override=True)

# 获取当前脚本所在的目录
current_directory = os.getcwd()

# 读取环境变量
api_token = os.getenv('API_TOKEN')

headers = {
    "Authorization": f"Bearer {api_token}",
    "Content-Type": "application/json"
}
def generateImage(model, prompt):
    body = {
        "inputs": translate_to_english(prompt)
    }
    if model == "pollinations-ai":
        r = requests.post("https://image.pollinations.ai/prompt/"+body['inputs'])
    else:
        r = requests.post("https://api-inference.huggingface.co/models/" + model,
                      data=json.dumps(body), headers=headers)
    # 将图片写入到 images 目录下，每个图片使用(时间戳+model).png 来命名
    timeStamp = str(int(time.time()))
    imagePath = "images/" + timeStamp + \
        "-" + model.split("/")[-1] + ".png"
    with open(imagePath, "wb") as f:
        f.write(r.content)
        f.close()
    voicePath = "voices/" + timeStamp + \
        "-" + model.split("/")[-1] + ".mp3"
    convert_text_to_speech(
        text=prompt, output_file=voicePath
    )


def convert_text_to_speech(text, output_file):
        # 指定输出目录
    output_directory = os.path.join(current_directory,"voices")
    # 创建输出目录（如果不存在）
    os.makedirs(output_directory, exist_ok=True)
    # 执行命令，并将工作目录设置为输出目录
    try:
        command = ['edge-tts', '--voice', 'zh-CN-XiaoyiNeural', '--text', text, '--write-media', output_file, '--write-subtitles', f'{output_file}.vtt']
        result = subprocess.run(command, cwd=current_directory, timeout=10)
        print(result)
        duration = get_duration_from_vtt(output_file + ".vtt")
        # 删除 无效音频 or 重新生成？
        if duration == 0.1:
            os.remove(output_file + ".vtt")
            os.remove(output_file)

    except subprocess.CalledProcessError as e:
        print("Command execution failed with return code:", e.returncode)
        print("Command output:", e.output)

def clear_folder(folder_path):
    """清空指定文件夹中的文件"""
    for filename in os.listdir(folder_path):
        file_path = os.path.join(folder_path, filename)
        if os.path.isfile(file_path):
            os.remove(file_path)



def split_sentences(text):
    pattern = r'[,.，。]'
    sentences = re.split(pattern, text)
    # 移除空白的句子
    sentences = [sentence.strip()
                 for sentence in sentences if sentence.strip()]
    return sentences
def convertTextToVideo(model, text):

    # 将文本段落进行分句
    sentences = split_sentences(text)

    # 清空 images 文件夹
    clear_folder("images")
    # 清空 voices 文件夹
    clear_folder("voices")

    # 为每个句子生成图片
    for sentence in sentences:
        if sentence.strip() != "":
            generateImage(model, sentence.strip())

    # 合成视频
    frame_width = 640
    frame_height = 480
    timeStamp = str(int(time.time()))
    output_video_path = "videos/" + timeStamp + \
        "-" + model.split("/")[-1] + ".mp4"
    output_video = cv2.VideoWriter(output_video_path, cv2.VideoWriter_fourcc(
        *'mp4v'), 30, (frame_width, frame_height))

    image_files = os.listdir('images')
    image_files.sort()

    for image_file in image_files:
        if image_file.endswith(".png"):

            text_color = (255, 255, 255)  # 白色文字
            background = (0, 0, 0,128)  # 黑色背景半透明
            image_path = "images/" + image_file
            draw_text = sentences[image_files.index(image_file)]
            add_text_to_image(draw_text, image_path,
                              text_color, background, padding=10)
            image = cv2.imread(image_path)
            resized_image = cv2.resize(image, (frame_width, frame_height))
            output_video.write(resized_image)
            # 添加停顿帧
            duration = get_duration_from_vtt(
                f"voices/{find_file_name_without_extension(image_file)}.mp3.vtt")
            print(duration)
            for _ in range(int(duration * 30)):
                output_video.write(resized_image)

    output_video.release()
    middle_output_video_path = "videos/" + timeStamp + \
        "-" + model.split("/")[-1] + ".withAudio.mp4"

    merge_audio_to_video("voices", output_video_path,
                         middle_output_video_path)
    desc_output_video_path = "videos/"+find_file_name_without_extension(
        middle_output_video_path)+"transformH264.mp4"
    convert_to_h264(middle_output_video_path, desc_output_video_path)
    return desc_output_video_path


def convert_to_h264(input_file, output_file):
    # 使用 FFmpeg 进行视频转换
    command = ['ffmpeg', '-i', input_file, '-c:v', 'libx264',
               '-preset', 'slow', '-crf', '22', '-c:a', 'copy', output_file]
    try:
        subprocess.run(command, check=True)
        print('视频转换成功！')
    except subprocess.CalledProcessError as e:
        print('视频转换失败:', e)



def find_file_name_without_extension(file_path):
    file_name = os.path.basename(file_path)
    file_name_without_extension = os.path.splitext(file_name)[0]
    return file_name_without_extension
def merge_audio_to_video(audio_directory, video_file, output_file):
    # 获取目录中的音频文件
    audio_files = [file for file in os.listdir(
        audio_directory) if file.endswith('.mp3')]

    if not audio_files:
        print("No audio files found in the directory.")
        return
    audio_files.sort()
    # 生成FFmpeg命令
    command = [
        'ffmpeg',
        '-i',
        video_file,
    ]

    # 添加音频文件参数
    for audio_file in audio_files:
        command.extend(['-i', audio_directory+'/'+audio_file])

    # 设置音频合并选项
    command.extend([
        '-filter_complex',
        ''.join([f'[{i+1}:0]' for i in range(len(audio_files))]) +
        f'concat=n={len(audio_files)}:v=0:a=1[outa]',
        '-map',
        '0:v',
        '-map',
        '[outa]',
        '-c:v',
        'copy',
        '-c:a',
        'aac',
        '-shortest',
        output_file
    ])

    # 执行FFmpeg命令
    result = subprocess.run(command,cwd=current_directory, timeout=300)
    print(result)

def get_duration_from_vtt(vtt_file):
    print(vtt_file)
    if not os.path.exists(vtt_file):
        return 0.1
    with open(vtt_file, 'r') as file:
        lines = file.readlines()

    total_duration = 0.1

    for line in lines:
        line = line.strip()
        if '-->' in line:
            start_time, end_time = line.split('-->')
            start_time = start_time.strip()
            end_time = end_time.strip()
            start_seconds = convert_time_to_seconds(start_time)
            end_seconds = convert_time_to_seconds(end_time)
            duration = end_seconds - start_seconds
            total_duration += duration

    return total_duration


def convert_time_to_seconds(time):
    hours, minutes, seconds = time.split(':')
    seconds, milliseconds = seconds.split('.')
    hours = int(hours)
    minutes = int(minutes)
    seconds = int(seconds)
    milliseconds = int(milliseconds)
    total_seconds = (hours * 3600) + (minutes * 60) + \
        seconds + (milliseconds / 1000)
    return total_seconds


if __name__ == '__main__':
   text_test= '''
   一个风和日丽的早上，我骑着自行车去学校，在路上遇到了彩虹，当时我的心情非常的愉快。
'''
#    convertTextToVideo(models[0], text_test)
